test/powershell/engine/Encoding.Tests.ps1

```diff
@@ -0,0 +1,225 @@
+            @{ Encoding = 'BigEndianUnicode'; Bytes = $testStringEncodedBytes['BigEndianUnicode'] },
+            @{ Encoding = 'Byte'; Bytes = $testStringEncodedBytes['Byte'] },
+            @{ Encoding = 'Default'; Bytes = $testStringEncodedBytes['Default'] },
+            # Oem encoding can change depending on system, calculate the expected string
+            @{ Encoding = 'Oem'; Bytes = $testStringEncodedBytes['Oem'] },
+            @{ Encoding = 'String'; Bytes = $testStringEncodedBytes['String'] },
+            @{ Encoding = 'UTF32'; Bytes = $testStringEncodedBytes['UTF32'] },
+            @{ Encoding = 'UTF7'; Bytes = $testStringEncodedBytes['UTF7'] },
+            @{ Encoding = 'UTF8'; Bytes = $testStringEncodedBytes['UTF8'] },
+            @{ Encoding = 'UTF8BOM'; Bytes = $testStringEncodedBytes['UTF8BOM'] },
+            @{ Encoding = 'UTF8NoBOM'; Bytes = $testStringEncodedBytes['UTF8NoBOM'] },
+            @{ Encoding = 'Unicode'; Bytes = $testStringEncodedBytes['Unicode'] },
+            @{ Encoding = 'Unknown'; Bytes = $testStringEncodedBytes['Unknown'] }
+
+    }
+
+    AfterEach {
+        if ( Test-Path $testFile )
+        {
+            remove-item $testFile
+        }
+        $PSDefaultFileEncoding = "Unknown"
+    }
+
+    It "Encoding for '<Encoding>' should have correct preamble '<preamble>'" -TestCase $preambleTests {
+        param ( $Encoding, $Preamble )
+        [Microsoft.PowerShell.PowerShellEncoding]::GetEncoding($Encoding).GetPreamble() -Join "-" | Should be $Preamble
+    }
+
+    It "Encoding for '<Encoding>' should create file with proper encoding" -TestCase $contentTests {
+        param ( $Encoding, $Bytes )
+        $testString | out-file -encoding $Encoding $testFile
+        Get-FileBytes $testFile | should be $Bytes
+    }
+
+    It "Setting PSDefaultFileEncoding to '<Encoding>' should create file with proper encoding" -TestCase $contentTests {
+        param ( $Encoding, $Bytes )
+        $PSDefaultFileEncoding = $Encoding
+        $testString | out-file $testFile
+        Get-FileBytes $testFile | should be $Bytes
+    }
+
+    It "Explicit encoding is not overridden by setting PSDefaultFileEncoding to '<Encoding>'" -TestCase $contentTests {
+        param ( $Encoding, $Bytes )
+        $PSDefaultFileEncoding = $Encoding
+        $testString | out-file -encoding ascii $testFile
+        Get-FileBytes $testFile | should be $testStringEncodedBytes['Ascii']
+    }
+
+    It "Explicit encoding set to unknown and preference variable set to unicode creates unicode file" {
+        $PSDefaultFileEncoding = "Unicode"
+        $testString | set-content -encoding unknown $testfile
+        Get-FileBytes $testFile | should be $testStringEncodedBytes['Unicode']
+    }
+
+    It "getting the encoding for an unknown cmdlet should return utf-8" {
+        $method = [microsoft.powershell.powershellencoding].getmember("GetWindowsLegacyEncoding","NonPublic,Static")
```

Comment:
We'd like to avoid reflection as much as possible in tests - can you find user accessible way to test this instead, e.g. calling a public api that calls this internal api?